astropy/modeling/parameters.py

```diff
@@ -504,9 +504,16 @@ def shape(self):
             model_axis = self._model._model_set_axis
 
             if model_axis < 0:
-                model_axis = len(shape) + model_axis
+                model_axis = len(self._model) + model_axis
 
-            shape = shape[:model_axis] + shape[model_axis + 1:]
+            # When a model set is initialized, the dimension of the parameters
+            # is increased by model_set_axis+1. To find the shape of a parameter
+            # within a single model the extra dimensions need to be removed first.
+            # The following dimension shows the number of models.
+            # The rest of the shape tuple represents the shape of the parameter
+            # in a single model.
+
+            shape = shape[model_axis + 1:]
```

Comment:
I think that makes more sense to me now, except that the model axis can go after the parameter shape, in which case I think this will always return an empty tuple (which is what I seem to see for `model_set_axis=2`)? But I think this is becoming a bit clearer.

Reply:
For a Polynomial, though, the coefficients should always have shape (), regardless of model_set_axis, right? At least that's my interpretation.

Reply:
I think so, but you **can** define one with a parameter that's a matrix. I have no idea what it means, but if I do it, I get this, for example:
```
> p2=models.Polynomial2D(2, c0_0=[[[0,1,2], [1,2,3]], [[3,4,5], [5,6,7]]], n_models=3, model_set_axis=2)
> p2.c0_0.value.shape
(2, 2, 3)
> p2.c0_1.value.shape
(1, 1, 3)
```
Does that give you any more insight, or is it just nonsense?

Reply:
I think this isn't a valid example but it doesn't fail. I'm not sure we have to put much effort into preventing such things.

Reply:
I suppose my point was that (whether or not it's a valid example), this perhaps illustrates how the shape was intended to work. Presumably a scalar parameter has a shape of `()` after removing the model set axis, while a 2x2 matrix parameter (where applicable) should have a shape of `(2,2)` etc. Based on my example, I think the above code will lose the shape in the latter case (ie. `c0_0`) for `model_set_axis=2`, won't it, even for some model where it makes sense? In fact, it should report different `c0_0` shapes for different `model_set_axis` values, which doesn't seem right (unless different models store the parameter values differently).

I can try to suggest something more specific if that helps...

Reply:
```python
> model_axis = 0
> sillyarg_ax0 = np.array([[[0,1],[1,2]], [[2,3],[3,4]], [[4,5],[5,6]]])
> p2 = models.Polynomial2D(2, c0_0=sillyarg_ax0, n_models=3, model_set_axis=model_axis)

> p2.c0_1.value.shape
(3,)
> p2.c0_1.value.shape[model_axis+1:]  # good
()
> p2.c0_0.value.shape
(3, 2, 2)
> p2.c0_0.value.shape[model_axis+1:]  # good
(2, 2)

> model_axis = 2
> sillyarg_ax2 = np.moveaxis(silly_ax0, 0, model_axis)
> p2 = models.Polynomial2D(2, c0_0=sillyarg_ax2, n_models=3, model_set_axis=model_axis)

> p2.c0_1.value.shape
(1, 1, 3)
> p2.c0_1.value.shape[model_axis+1:]  # good
()
> p2.c0_0.value.shape
(2, 2, 3)
> p2.c0_0.value.shape[model_axis+1:]  # bad
()
```
No? Here I've used `p2.c0_0.value.shape` instead of `p2._param_metrics['c0_0']['shape']`, but I think they're the same thing?

Reply:
Yes, you are right. That's why I have the change to the test here

https://github.com/astropy/astropy/pull/7222/files#diff076341e0d057b9f52a7b5946df2a3453R57

There are other cases when this won't work but I think they all happen when parameters are stacked in an array along any axis that is not the 0th axis. 

I have no solution for this other than document that in this case the parameters should be along that specific axis. How much of a burden/inconvenience would this be, do you think?

I'm also open to other suggestions/ideas.

Reply:
Are you asking how inconvenient it would be if parameters can only accept non-scalar values (per model) when `model_set_axis=0` and not for higher model set axes? I think, for me, the answer is currently no problem at all, since I don't have an application for non-scalar parameters, or even really understand their purpose, beyond passing arrays to affine transform. I was just trying to avoid breaking existing code/assumptions. It does bothers me slightly when results seem inconsistent with the intentions though :slightly_frowning_face:.

I was trying to understand the purpose of all that pivot stuff in `_prepare_inputs_model_set`... Indeed, the pivot dimension is where `model_set_axis` gets moved to during model evaluation (which I think must be what you were saying above), but why? It seems like maybe the idea is to leave the right number of trailing dimensions after `model_set_axis` to ensure that the input arrays can be broadcast with non-scalar parameters?? But I think model *parameters* always include the `model_set_axis` in its nominal position and if you keep it at that same dimension for the inputs too, everything else should broadcast together in just the same way as it would for a single model, whatever the axis? So I can't see why pivot isn't just set to `model.model_set_axis` at L3341 and L3354 (which actually worked in the case I tried), the same as in the other cases, but I may just be thoroughly confused... (and this still doesn't address shape being inconsistent with its comment). Anyway, I should probably get back to other things.